### PR TITLE
Add multi auth feature to backend

### LINKED ISSUE #3316: Enable logging in with either UAA or Github with the ability to turn off Github independently.
#### User Story

A Federalist user has the option to login via UAA or GitHub

#### Background (Optional)

<TODO>

#### Acceptance Criteria

- [ ] Users with UAA identity can only log in via UAA and not via GitHub
- [ ] Explain to users how to add GitHub content if UAA is enabled.
- [ ] Alert GitHub only authenticated users that GitHub login method will be removed.
- [ ] Change made live via deploy of Federalist

_After evaluating, edit this part:_

#### Implementation outline (if higher than "low" effort):

- [ ] <TODO>

### PR DESCRIPTION
**Pull request by apburnes**
Closes #3316 

#### Todo:
- [x] Users with UAA identity can only log in via UAA and not via GitHub
- [x] Explain to users how to add GitHub content if UAA is enabled.
- [x] Alert GitHub only authenticated users that GitHub login method will be removed.
- [x] Change made live via deploy of Federalist
- [x] Add feature flag to deployment for the multi auth
- [x]  Update from `upsert` to `update` for github auth

### CODE REVIEW

File: views/navigation.njk [hunk 1]
```diff
@@ -38,12 +38,28 @@
             </a>
           </li>
           {% if isAuthenticated %}
-          <li class="nav-action">
-            <a class="usa-button usa-button-white" href="/logout">Log out</a>
-          </li>
+            <li class="nav-action">
```
Comment (davemcorwin) [inline views/navigation.njk:41]: does the `isUAA` make sense anymore given that it's no longer binary?

File: views/home.njk [hunk 2]
```diff
@@ -12,15 +12,37 @@
...
+                  <button class="usa-button">
+                    Agree and continue with cloud.gov
+                  </button>
+                </a>
+              </div>
+            {% else %}
+              {% if isUaa %}
+                <a href="/login">
+                  <button class="usa-button">
+                    Agree and continue with cloud.gov's SecureAuth
```
Comment (davemcorwin) [inline views/home.njk:35]:
```suggestion
                    Agree and continue with cloud.gov
```

File: api/services/passport.js [hunk 3]
```diff
@@ -18,17 +22,34 @@ const {
   uaa: { options: uaaOptions },
 } = config.passport;
 
+// eslint-disable-next-line consistent-return
+async function checkMultiAuth(username, callback) {
```
[the remark this answers is not in the document]
Comment (apburnes) [reply api/services/passport.js:26]: Do we want to block all new user creation or make us run the invite/org creation process based on demand while  transition folks to UAA? That timeline will likely stretch a few months and our new, unrefined UAA/Org/Onboarding process could become the blocker for adding new users to the platform.  Associating existing Github users (including new accounts during transition) to UAA is straight forward and allows users to have continued autonomy while giving us time to refine the new process.

[the remark this answers is not in the document]
Comment (davemcorwin) [reply api/services/passport.js:26]: With the recent PR merge, Pages Operators have the ability to invite users outside of the org creation process, this includes migrating an existing users. Once this goes live (after SCR, end of July), I don't think we should allow new user creation outside of UAA. My understanding is that we only have until beginning of September to get everyone moved over.  

We will still need improvements to this process and the ability to manage users/orgs.

File: api/routers/auth.js [hunk 4]
```diff
@@ -37,7 +35,7 @@ router.get('/auth/uaa/logout', (_req, res) => res.redirect('/'));
 const onGithubSuccess = async (req, res) => {
   const githubUser = req.user;
 
-  const user = await User.findByPk(req.session.passport.user);
+  const user = await User.findByPk(req.session.passport.user.id);
```
[the remark this answers is not in the document]
Comment (apburnes) [reply api/routers/auth.js:38]: Good catch. I tested creating the serialized user into an object with `id` & `authtype` and missed this in cleanup.

File: views/navigation.njk (hunk 1 above)
Comment (apburnes) [reply views/navigation.njk:41]: I think we'll need it until we turn on the multi auth flag and possibly for staging until we move completely to UAA.

File: views/home.njk [hunk 5]
```diff
@@ -12,15 +12,37 @@
...
-            <a href="/login">
-              <button class="usa-button">
-                {% if isUAA %}
-                  Agree and continue with SecureAuth
-                  {% else %}
-                  <img src="/images/github-logo.png"> Agree and continue with GitHub
-                {% endif %}
-              </button>
-            </a>
+            {% if hasMultiAuth %}
```
[the remark this answers is not in the document]
Comment (apburnes) [reply views/home.njk:15]: Updated in #3408

File: api/services/passport.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (apburnes) [reply api/services/passport.js]: Updated to  user `update` from `upsert`.

File: api/controllers/main.js [hunk 6]
```diff
@@ -18,11 +19,24 @@ function defaultContext(req) {
     homepageUrl: config.app.homepageUrl,
     webpackAssets,
     isUAA: config.env.authIDP === 'uaa',
```
[the remark this answers is not in the document]
Comment (apburnes) [reply api/controllers/main.js:21]: Switched to `FEATURE_AUTH_UAA` & `FEATURE_AUTH_GITHUB`

File: views/navigation.njk (hunk 1 above)
Comment (apburnes) [reply views/navigation.njk:41]: Same as above

**Comment by apburnes** [comment]
Ready for review again.  Checkout the following screenshots for the multiple auth type states.

When both UAA and Github auth are enabled.
Screen Shot 2021-05-28 at 1 45 28 PM

When both UAA and Github auth are enabled but you are logged in with Github without a UAA identity yet.
Screen Shot 2021-05-28 at 1 46 39 PM

When just Github auth is enabled.
Screen Shot 2021-05-28 at 1 49 15 PM

When just UAA auth is enabled.
Screen Shot 2021-05-28 at 1 51 37 PM

File: views/home.njk [hunk 7]
```diff
@@ -12,20 +12,41 @@
...
-              <button class="usa-button">
-                {% if isUAA %}
-                  Agree and continue with cloud.gov authentication
-                  {% else %}
-                  <img src="/images/github-logo.png"> Agree and continue with GitHub
+            {% if authUAA %}
+              <div class="well-gray-lightest">
+                {% if authGithub %}
+                  <h5>Did we invite to use your cloud.gov account?</h5>
                 {% endif %}
```
Comment (davemcorwin) [inline views/home.njk:19]:
```suggestion
```

Comment (davemcorwin) [reply views/home.njk:19]: maybe leave this as the happy path and only have the disclaimer with the github auth below?

File: views/home.njk [hunk 8]
```diff
@@ -12,20 +12,41 @@
...
+              <div class="well">
+                <p style="text-align: center;">
+                  or...
+                </p>
+              </div>
+            {% endif %}
+            {% if authGithub %}
+              <div class="well-gray-lightest">
+                {% if authUAA %}
+                  <h5>Waiting on your cloud.gov account? Use your Github account.</h5>
```
Comment (davemcorwin) [inline views/home.njk:37]:
```suggestion
                  <h5>Don't have cloud.gov account yet? Use your Github account.</h5>
```

File: views/navigation.njk [hunk 9]
```diff
@@ -38,12 +38,20 @@
...
+                <a class="usa-button usa-button-white" href="/logout/github">Log out</a>
+              {% endif %}
+            </li>
           {% else %}
             <li  class="nav-action">
-              <a href="/login" class="usa-button usa-button-white">Log in</a>
+              {% if authUAA %}
+                <a href="/login" class="usa-button usa-button-white">Log in with cloud.gov</a>
+              {% else %}
+                <a href="/login/github" class="usa-button usa-button-white">Log in Github</a>
```
Comment (davemcorwin) [inline views/navigation.njk:53]:
```suggestion
                <a href="/login/github" class="usa-button usa-button-white">Log in with Github</a>
```

**Review by davemcorwin — APPROVED** [review]
Thanks @apburnes ! Looks great, made a couple small content suggestions.

1) if deployed as is to staging, nothing should change, right? (except for maybe content)
2) Are we still using the `AUTH_IDP` env var?

File: views/home.njk (hunk 7 above)
Comment (apburnes) [reply views/home.njk:19]: 👍

**Comment by apburnes** [comment]
@davemcorwin 
> if deployed as is to staging, nothing should change, right? (except for maybe content)
- Yes, will deploy with UAA auth for staging and Github auth in production. Only some content and the light grey/blue well around the authentication button

> Are we still using the AUTH_IDP env var?
- Let me verify.

**Pull request merged**